Modified LSTMPolicy to have instance variables for feedforward activations

kantneel opened the pull request:
This is the cleanest way to make activations accessible to a wrapper which wishes to exposes them.

kantneel committed:
added TransparentLSTMPolicy

kantneel committed:
removed Transparent things since they're in adversarial-policies now

kantneel committed:
modified master version of LSTMPolicy to have instance variables for feedforward activations

kantneel committed:
consolidated ff_outs into one list, similar to state_outs

kantneel committed:
made hiddens an instance variable in LSTMPolicy to give access in methods other than __init__

AdamGleave commented on gym_compete/policy.py:
```diff
@@ -155,9 +155,10 @@ def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, hiddens=No
                          scope=scope, reuse=reuse, normalize=normalize)
         if hiddens is None:
             hiddens = [128, 128]
+        self.hiddens = hiddens
```
Why does this need to be an instance variable? (Don't particularly object to it, just not sure I understand why it needs to be.)

AdamGleave commented on gym_compete/policy.py:
```diff
@@ -212,19 +213,20 @@ def loop_fn(time, cell_output, cell_state, loop_state):
                     self.state_out.append(final_state)
 
-                    return last_out
+                    return last_out, input_seq
```
Is `input_seq` really what we want? It's the output of the last hidden layer before going into the LSTM. If we're going to pick just one layer it seems like a reasonable one to pick, but why not expose all the layers so we can choose which one to look at during experiments? This would also be needed to support some more advanced methods like the HER idea where we'd start at the first layer and work up to deeper layers over time.

AdamGleave commented on gym_compete/policy.py:
```diff
@@ -212,19 +213,20 @@ def loop_fn(time, cell_output, cell_state, loop_state):
+                    return last_out, input_seq
 
-                value_out = lstm(0, 'v')
+                value_out, value_ff_acts = lstm(0, 'v')
```
I think we're unlikely to want to use value activations since they don't directly determine the action taken. But no harm in exposing it.

AdamGleave commented on gym_compete/policy.py:
```diff
@@ -212,19 +213,20 @@ def loop_fn(time, cell_output, cell_state, loop_state):
                 logstd = tf.get_variable(name="logstd", shape=[1, ac_space.shape[0]],
                                          initializer=tf.zeros_initializer())
 
+                self.ff_out = [value_ff_acts, policy_ff_acts]
```
Think it'd be good to have difference instance variables (or a dict with keys) for each of value and policy activations, as as we often might only want to look at one of them.

AdamGleave reviewed (changes requested):
If we're going to make this change, can we update MLPPolicy as well? I'd like them to have a consistent interface. It might be worth pushing some of the logic up to the base `GymCompetePolicy` class as well. e.g. If we want to make hiddens an instance variable, maybe that should be set in `GymCompetePolicy.__init__`

kantneel committed:
added self.hiddens for MlpPolicyValue and refactored activation exposure for LSTMPolicy

kantneel commented:
I made the change for MlpPolicyValue too. GymCompetePolicy currently doesn't accept hiddens as a constructor argument, and hiddens' default value is None in both MlpPolicyValue and LSTMPolicy, where its final value is handled after the super constructor call. For these reasons I didn't feel it necessary to make hiddens an instance variable of GymCompetePolicy. But if you'd like I can refactor it accordingly.

AdamGleave commented:
Where are the activations exposed for `MlpPolicyValue`? I do not see any `ff_outs` instance variable. 

What shape are the activations supposed to be? The `(self.n_env, n_steps, ) + ob_space` thing only makes sense for LSTM policies; in MLP policies the activations will just be `(n_batch, ) + ob_space`. Maybe it's OK they're different shapes? You have a better ideas as you're writing the code that uses this.

I think I'd prefer to add `hiddens` to `GymCompetePolicy` (it's only ever used in this file, it's a class I added, so OK to break the API) and handle setting the `hiddens` default value before the super constructor call. (This isn't Java, it's OK to do stuff before super constructors.)

kantneel committed:
created self.ff_out for MlpPolicyValue

kantneel committed:
self.hiddens instance variable for GymCompetePolicy

Commit:
Merge branch 'master' into transparent

kantneel commented:
@AdamGleave I addressed your comments. Please review when you get a chance. Once this is approved I can run tests on the `transparent` branch of adversarial-policies to ensure that `TransparentMlpPolicyValue` works correctly.

AdamGleave reviewed (approved):
LGTM
